torch/nn/modules/module.py

```diff
@@ -823,26 +825,47 @@ def compute_should_use_set_data(tensor, tensor_applied):
             # `with torch.no_grad():`
             with torch.no_grad():
                 param_applied = fn(param)
-            should_use_set_data = compute_should_use_set_data(param, param_applied)
-            if should_use_set_data:
-                param.data = param_applied
-                out_param = param
+            p_should_use_set_data = compute_should_use_set_data(param, param_applied)
+            param_grad = param.grad
+            if p_should_use_set_data:
+                if should_use_swap_tensors:
+                    if param_grad is not None:
+                        # Accessing param.grad makes its use_count 2, which will prevent swapping.
+                        # Decrement use count of the gradient by setting to None
+                        param.grad = None
+                    param_applied = torch.nn.Parameter(param_applied, requires_grad=param.requires_grad)
+                    try:
+                        torch.utils.swap_tensors(param, param_applied)
+                    except Exception as e:
+                        raise RuntimeError(f"Couldn't swap {key}") from e
```

Comment:
nit: should this be more precise about while moving module parameter or something? The key here will only contain the key within this module, not the fqn!